File: WordDistanceFinder.java
/* This class will be given a list of words (such as might be tokenized
 * from a paragraph of text), and will provide a method that takes two
 * words and returns the shortest distance (in words) between those two
 * words in the provided text.
 * Example:
 *   WordDistanceFinder finder = new WordDistanceFinder(Arrays.asList("the", "quick", "brown", "fox", "quick"));
 *   assert(finder.distance("fox","the") == 3);
 *   assert(finder.distance("quick", "fox") == 1);
 *
 * "quick" appears twice in the input. There are two possible distance values for "quick" and "fox":
 *     (3 - 1) = 2 and (4 - 3) = 1.
 * Since we have to return the shortest distance between the two words we return 1.
 */
public class WordDistanceFinder {
    public WordDistanceFinder (List<String> words) {
        MultiMap<String, Integer> wordMap = new MultiMap<>();
        for (int index = 0; index < words.size(); index++) {
        	wordMap.put(words[index], index);
        }
    }
    public int distance (String wordOne, String wordTwo) {
        LinkedList<Integer> wordOneIndices = wordMap.get(wordOne);
        LinkedList<Integer> wordTwoIndices = wordMap.get(wordTwo);

        int wordOneIndex = wordOneIndices.remove();
        int wordTwoIndex = wordTwoIndices.remove();
        int answer;

        while (!wordOneIndices.isEmpty() && !wordTwoIndices.isEmpty) {

        }
    }
}

File: WordDistanceFinder.py
class WordDistanceFinder:
    """
    This class will be given a list of words and will provide a method that
    takes two words and returns the shortest distance (in words) between 
    those two words in the provided text.

    Example:
        finder = WordDistanceFinder(["the", "quick", "brown", "fox", "quick"])
        finder.distance("fox", "the") -> 3
        finder.distance("quick", "fox") -> 1
    """

    def __init__(self, words):
        pass

    def distance(self, wordOne, wordTwo):
        pass

    ## MY SOLUTION
    # def __init__(self, words):
    #     self.word_map = dict()
    #     for index in xrange(len(words)):
    #         word = words[index]
    #         if word in self.word_map:
    #             self.word_map[word] += [index]
    #         else:
    #             self.word_map[word] = [index]
         
    # def distance(self, wordOne, wordTwo):
    #     min_distance = -1
    #     if not(wordOne in self.word_map.keys() and wordTwo in self.word_map.keys()):
    #         return min_distance
    #     wordOne_indices = self.word_map[wordOne]
    #     wordTwo_indices = self.word_map[wordTwo]

    #     for wordOne_index in wordOne_indices:
    #         for wordTwo_index in wordTwo_indices:
    #             distance = abs(wordTwo_index - wordOne_index)
    #             if (min_distance == -1) or (distance < min_distance):
    #                 min_distance = distance

    #     return min_distance
